common/src/main/java/org/cloudfoundry/identity/uaa/authorization/UaaAuthorizationEndpoint.java

```diff
@@ -418,24 +430,20 @@ private String getUnsuccessfulRedirect(AuthorizationRequest authorizationRequest
         }
 
         UriComponentsBuilder template = UriComponentsBuilder.fromUriString(authorizationRequest.getRedirectUri());
-        Map<String, String> query = new LinkedHashMap<String, String>();
+        //decode already encoded values
+        template = UriComponentsBuilder.fromHttpUrl(template.build(true).toUriString());
         StringBuilder values = new StringBuilder();
 
-        values.append("error={error}");
-        query.put("error", failure.getOAuth2ErrorCode());
-
-        values.append("&error_description={error_description}");
-        query.put("error_description", failure.getMessage());
+        values.append("error="+encode(failure.getOAuth2ErrorCode()));
+        values.append("&error_description="+encode(failure.getMessage()));
```

Comment:
Each variable that we add, we must encode, here we know that the parameter names do not need encoding